Create super-class OperationalDatasetView out of OperationalDataset

ksperling-apple opened the pull request:
#### Summary

OperationalDatasetView does not copy or hold a buffer for the underlying dataset, and provides all the read-only methods that were present on OperationalDataset.  OperationalDataset has its own buffer and copies the dataset as before.

Also simplify some of the mechanics of updating values in the dataset, by replacing MakeRoom with InsertOrReplace that directly takes the required size of the value and handles updating both the tag and length of the TLV item as well as of the overall dataset, so the caller only has to fill in the value.

#### Testing

Existing TestOperationalDataset tests still pass, plus some new tests for both OperationalDatasetView and OperationalDataset.

ksperling-apple committed:
Create super-class OperationalDatasetView out of OperationalDataset

OperationalDatasetView does not copy or hold a buffer for the underlying
dataset, and provides all the read-only methods that were present on
OperationalDataset.  OperationalDataset has its own buffer and copies the
dataset as before.

Also simplify some of the mechanics of updating values in the dataset, by
replacing MakeRoom with InsertOrReplace that directly takes the required size
of the value and handles updating both the tag and length of the TLV item as
well as of the overall dataset, so the caller only has to fill in the value.

ksperling-apple commented on src/lib/support/ThreadOperationalDataset.cpp:
```diff
@@ -59,80 +59,77 @@ class ThreadTLV final
+    void SetValue(const void * aValue, size_t aLength)
     {
-        SetLength(aLength);
+        assert(GetLength() == aLength);
         memcpy(GetValue(), aValue, aLength);
```
There is no such API contract.

ksperling-apple committed:
Avoid out of bounds access in IsValid

ksperling-apple commented:
/gemini review

ksperling-apple committed:
use EXPECT_STREQ instead of strncmp in test

ksperling-apple committed:
Handle copy construction / assignment correctly

ksperling-apple committed:
Fix test warning and use the view in TND server

bzbarsky-apple commented on src/lib/support/ThreadOperationalDataset.cpp:
```diff
@@ -38,11 +38,11 @@ namespace Thread {
-public:
-    static constexpr uint8_t kMaxLength = kLengthEscape - 1;
+    static constexpr uint8_t kHeaderSize = 2;      ///< Size of the TLV header (Type + Length).
+    static constexpr uint8_t kMaxLength  = kSizeOperationalDataset - kHeaderSize;
```
I'm a little confused by these length/size changes.  Below, we allow SetLength to set a value greater than this `kMaxLength` constant, right?  Is that expected?

ksperling-apple replied:
I think the old constant was simply wrong, since an individual TLV element can't be longer than the whole dataset. SetLength is just enforcing "we don't support multi-byte lengths", I changed it from `!=` to `<` because the argument changed from uint_8 to size_t. I guess I could check for `<= maxLength` instead.

This ThreadTLV type is just an internal helper, so it's somewhat arbitrary what's validated here vs in the OperationalDataset logic.

bzbarsky-apple commented on src/lib/support/ThreadOperationalDataset.cpp:
```diff
@@ -142,58 +139,56 @@ class ThreadTLV final
+    while (tlv != end)
     {
-        return CHIP_ERROR_INVALID_ARGUMENT;
+        VerifyOrReturnValue(tlv + 1 <= end, false);                               // out of bounds
```
The old IsValid looks like it did a `<` test here (i.e. disallowed things with `GetLength() == 0`).  Is the behavior change here intentional?

ksperling-apple replied:
The old logic was
```
while (curr + sizeof(ThreadTLV) < end) { ... }
return curr == end;
```
So an empty dataset is valid. This also make sense in terms of the invariant that `IsValid(mData)` remains true, and we start with an empty dataset.

bzbarsky-apple commented on src/lib/support/ThreadOperationalDataset.cpp:
```diff
@@ -287,240 +252,233 @@ CHIP_ERROR OperationalDataset::GetMasterKey(uint8_t (&aMasterKey)[kSizeMasterKey
+    {
+        size_t tlvLength = tlv->GetLength();
+        VerifyOrReturnValue(aValueSize != tlvLength, tlv); // re-use in place if same size
+        VerifyOrReturnValue(aValueSize < tlvLength || mData.size() + aValueSize - tlvLength <= sizeof(mBuffer), nullptr);
```
Depending on how much we trust our inputs, this might be a more overflow-safe check:
```suggestion
        VerifyOrReturnValue(aValueSize < tlvLength || aValueSize <= sizeof(mBuffer) - (mData.size() - tlvLength), nullptr);
```

ksperling-apple replied:
This is an internal helper methods and the caller is expected to have already validated aValueSize makes sense for the specific value being set, or checked against the max size (kMaxDatasetElementLength) if there is no other constraint. I've added an `assert` and a comment, so I think we can leave these expressions in the maybe more readable arrangement.

bzbarsky-apple commented on src/lib/support/ThreadOperationalDataset.cpp:
```diff
@@ -287,240 +252,233 @@ CHIP_ERROR OperationalDataset::GetMasterKey(uint8_t (&aMasterKey)[kSizeMasterKey
+    }
+    else
+    {
+        VerifyOrReturnValue(mData.size() + sizeof(ThreadTLV) + aValueSize <= sizeof(mBuffer), nullptr);
```
This one is a bit harder to make overflow-safe, if it's needed...  Would need something like:
```suggestion
        VerifyOrReturnValue(mData.size() + sizeof(ThreadTLV) <= sizeof(mBuffer) && aValueSize <= sizeof(mBuffer) - (mData.size() + sizeof(ThreadTLV)), nullptr);
```

ksperling-apple replied:
As above

bzbarsky-apple commented on src/lib/support/ThreadOperationalDataset.h:
```diff
@@ -270,88 +191,219 @@ class OperationalDataset
 
     /**
-     * This method sets the delay timer within the dataset.
+     * This method sets Thread active timestamp to the dataset.
```
"to" is odd here.  "within" or "in" would be better?  Same for the other setters.  I know this is pre-existing; followup is fine.

ksperling-apple replied:
Did a pass over these descriptions, should be worded better now.

bzbarsky-apple commented on src/lib/support/ThreadOperationalDataset.h:
```diff
@@ -270,88 +191,219 @@ class OperationalDataset
+     * Returns a ByteSpan view of the current state of the dataset.
+     * The byte span is only valid as long the OperationalDataset object is alive and not modified.
+     */
+    ByteSpan AsByteSpan() const { return this->OperationalDatasetView::AsByteSpan(); }
```
Why is this needed, if we publicly inherit from OperationalDatasetView?  I guess the documentation about invariants is different?

ksperling-apple replied:
Yes this is just so I had somewhere to put the comment, not sure if there's a more elegant way to do that.

ksperling-apple committed:
Address review comments

- Replace ThreadTLV::kMaxLength with kMaxDatasetElementLength outside of it,
  the TheadTLV helper does not need to know about kSizeOperationalDataset.
- assert(aValueSize <= kMaxDatasetElementLength) in InsertOrReplace and
  document that callers (the various Set... methods) are responsible for
  checking the value size.
- Check Channel Page == 0 in GetChannel and use BigEndian instead of math.
- Tidy up method descriptions in the header.

andy31415 commented on src/lib/support/ThreadOperationalDataset.h:
```diff
@@ -133,225 +127,289 @@ class OperationalDataset
+    static bool IsValid(ByteSpan aData);
+
+private:
+    friend class OperationalDataset;
```
why friend? Can we do protected and not use fried?

andy31415 commented on src/lib/support/ThreadOperationalDataset.h:
```diff
@@ -133,225 +127,289 @@ class OperationalDataset
+ * This class provides methods to manipulate a Thread operational dataset.
+ * It maintains an internal buffer sized to accommodate the maximum possible dataset size.
+ */
+class OperationalDataset : public OperationalDatasetView
```
Did you consider splitting h/cpp so that each class has its own h/cpp file? Why don't we have ThreadOperationalDataSetView.h/cpp?

ksperling-apple committed:
Expand comment on self-assignments
